node/src/peer_tracker.rs

```diff
@@ -1,56 +1,163 @@
 use std::borrow::Borrow;
 
-use dashmap::DashSet;
-use libp2p::PeerId;
+use dashmap::mapref::entry::Entry;
+use dashmap::mapref::one::RefMut;
+use dashmap::DashMap;
+use libp2p::{identify, Multiaddr, PeerId};
 
+#[derive(Debug)]
 pub struct PeerTracker {
-    peers: DashSet<PeerId>,
+    peers: DashMap<PeerId, PeerInfo>,
+}
+
+#[derive(Debug)]
+struct PeerInfo {
+    addrs: Vec<Multiaddr>,
+    state: PeerState,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum PeerState {
+    Discovered,
+    AddressesFound,
+    Connected,
+    Identified,
+}
+
+impl PeerInfo {
+    fn is_connected(&self) -> bool {
+        matches!(self.state, PeerState::Connected | PeerState::Identified)
+    }
 }
 
 impl PeerTracker {
     pub fn new() -> Self {
         PeerTracker {
-            peers: DashSet::new(),
+            peers: DashMap::new(),
         }
     }
 
-    pub fn add(&self, peer: PeerId) {
-        self.peers.insert(peer);
+    pub fn maybe_discovered(&self, peer: PeerId) -> bool {
+        match self.peers.entry(peer) {
+            Entry::Vacant(entry) => {
+                entry.insert(PeerInfo {
+                    addrs: Vec::new(),
+                    state: PeerState::Discovered,
+                });
+                true
+            }
+            Entry::Occupied(_) => false,
+        }
     }
 
-    pub fn add_many<I, P>(&self, peers: I)
+    /// Get the `PeerInfo` of the peer.
+    ///
+    /// If peer is not found it is added as `PeerState::Discovered`.
+    fn get(&self, peer: PeerId) -> RefMut<PeerId, PeerInfo> {
+        self.peers.entry(peer).or_insert_with(|| PeerInfo {
+            addrs: Vec::new(),
+            state: PeerState::Discovered,
+        })
+    }
+
+    pub fn add_addresses<I, A>(&self, peer: PeerId, addrs: I)
     where
-        I: IntoIterator<Item = P>,
-        P: Borrow<PeerId>,
+        I: IntoIterator<Item = A>,
+        A: Borrow<Multiaddr>,
     {
-        for peer in peers {
-            self.add(*peer.borrow());
+        let mut state = self.get(peer);
+
+        for addr in addrs {
+            let addr = addr.borrow();
+
+            if !state.addrs.contains(addr) {
+                state.addrs.push(addr.to_owned());
+            }
+        }
+
+        // Upgrade state
+        if state.state == PeerState::Discovered && !state.addrs.is_empty() {
+            state.state = PeerState::AddressesFound;
         }
     }
 
-    pub fn remove(&self, peer: PeerId) {
-        self.peers.remove(&peer);
+    pub fn clear_addresses(&self, peer: PeerId) {
+        let mut state = self.get(peer);
+
+        state.addrs.clear();
+        state.addrs.shrink_to_fit();
+
+        // Downgrade state. Note that we do not downgrade a connected peer.
```

Comment:
Is it even okay to remove the addresses of connected peer, or maybe the one that we are connected through?

Reply:
Not sure. I don't use this method yet

Reply:
I also had an idea that the connected address can be stored in a different field as a primary address. But in general I'm not sure if we really need to keep track of the addresses